### DISCUSSION TITLE
Using PicoW as mass storage device

### DISCUSSION THREAD

**Discussion by Deemocean** [earlephilhower/arduino-pico#761] [Q&A]
It seems like the Arduino-Pico File system(SdFat ver 2) is not compatible with the `Adafruit fork SdFat`--which is essential for the TinyUSB MSC implementation(`readBlock()`, `writeBlock()`,  etc.). #749
In my project, I am thinking about using WiFi.h and TinyUSB, making it possible to send files through wifi to the SD card on the Pico W, and letting the file be readable through the device it is connecting to (a 3d printer) through the USB. 
So I was wondering, is there a way around this SdFat problem, or is this just a dead-end for now?
Thanks for any help in advance!

**Comment by earlephilhower** [comment]
I'm not quite following the issue here, to be honest.

To make a USB mass storage device, you don't need *any* filesystem support on the Pico.  A USB MSD just reads and writes LBAs requested by the host.

To share the files uploaded over Wifi, you *do* need a filesystem and a web server.  The web server, you can probably get from an Arduino library.  The filesystem, you can use the SD/SDFS filesytem (and SdFat 2 underneath it) to read and write files to send over the WiFi.

**Reply by Deemocean** [reply]
Sorry if I am asking newbie questions...
I was not really sure how to make Pico act as an MSD, so I just followed the examples given by the TinyUSB library.
Which requires me to set 3 callback functions 
```C++
// Set read-write callback
usb_msc.setReadWriteCallback(msc_read_cb, msc_write_cb, msc_flush_cb);
```
They were somewhat black boxes for me
Taking `msc_read_cb` as an example:
```C++
// Callback invoked when received READ10 command.
// Copy disk's data to buffer (up to bufsize) and
// return number of copied bytes (must be multiple of block size)
int32_t msc_read_cb(uint32_t lba, void *buffer, uint32_t bufsize)
{
  (void)bufsize;
  return card.readBlock(lba, (uint8_t *)buffer) ? 512 : -1;
}
```
This `readBlock()` function is only implemented by the Adafruit fork SdFat, and I guess this is the only place the filesystem support got related.
_Please correct me if I am wrong:_
TinyUSB takes care of the interaction between Pico and the device connected(as a driver?), and what I should do is **implement those callback functions?**

**Really appreciated your response! Thanks!**

**Reply by earlephilhower** [reply]
Ah, they are *not* using the FS in this mode, just the card read/write operation.  SDFat 2.0 was quite a refactoring so the call is still probably there, but not in the same object or the same function name.  A quick look through the headers should get you started.

>  TinyUSB takes care of the interaction between Pico and the device connected(as a driver?), and what I should do is implement those callback functions?

Yes, that seems like what is needed to make a MSD USB device.  I've not looked at it deeply so please don't take my word as gospel, yet!
